index.js

```diff
@@ -20,39 +20,38 @@
 //. purpose: the [`typeof`][1] operator and [`Object.prototype.toString`][2].
 //. Each has pros and cons, but neither supports user-defined types.
 //.
-//. This package specifies an [algorithm][3] for deriving a _type identifier_
-//. from any JavaScript value, and exports an implementation of the algorithm.
-//. Authors of algebraic data types may follow this specification in order to
-//. make their data types compatible with the algorithm.
+//. This package exports a function which derives a _type identifier_ from any
+//. JavaScript value, and a [specification][3] for customizing the type
+//. identifier of a value.
 //.
-//. ### Algorithm
+//. ### Specification
 //.
-//. 1.  Take any JavaScript value `x`.
+//. For a type to be compatible with the algorithm:
 //.
-//. 2.  If `x` is `null` or `undefined`, go to step 6.
-//.
-//. 3.  If `x.constructor` evaluates to `null` or `undefined`, go to step 6.
-//.
-//. 4.  If `x.constructor.prototype === x`, go to step 6. This check prevents a
-//.     prototype object from being considered a member of its associated type.
+//.   - every member of the type MUST have a `constructor` property pointing
+//.     to an object known as the _type representative_;
 //.
-//. 5.  If `typeof x.constructor['@@type']` evaluates to `'string'`, return
-//.     the value of `x.constructor['@@type']`.
+//.   - the type representative MUST have a `@@type` property;
 //.
-//. 6.  Return the [`Object.prototype.toString`][2] representation of `x`
-//.     without the leading `'[object '` and trailing `']'`.
+//.   - the type representative's `@@type` property (the _type identifier_)
+//.     MUST be a string primitive and SHOULD have format:
+//.     `'<namespace>/<name>[@<version>]'`, where:
 //.
-//. ### Compatibility
+//.       - The `namespace` MUST consist of one or more non-`/` characters, MAY
+//.         contain a single `/` if it starts with `@` (to support
+//.         [npm scopes][4]), and SHOULD equal the npm package name which
+//.         defines the type; and
```

Comment:
> I could modify the expression to consider this specific case invalid, but then `scope@home/Type@1` would still be considered valid, since we are allowing for URL-unsafe characters (unlike npm).

This highlights another thought I have not conveyed clearly. My goal is not for every valid `namespace` be a valid npm package name, but for every valid npm package name to be a valid `namespace`. I believe we should make allowances for npm package names, but not require them. A type defined in a private package, for example, has no public package name to reference in its type identifier.

> Either we allow `@scope` to be a valid namespace (with the exception being how the first out of multiple slashes is treated, as implied by my specification); *or* we validate the namespace part properly, which did not have your preference.

We can have our :cake: and eat it too, I think. In https://github.com/sanctuary-js/sanctuary-type-identifiers/pull/8#issuecomment-294476090 I proposed this grammar for `namespace`:

```bnf
<namespace> ::= @ <scope> / <package> | <package>
```

`xyz` and `@abc/xyz` should be valid; `@abc` should be invalid.